### ISSUE TITLE
mpld3 json array dumping on python=3, numpy=1.14

### ISSUE THREAD

**Issue by David-Baddeley** [python-microscopy/python-microscopy#95]
**[Original report](https://bitbucket.org/david_baddeley/python-microscopy/issue/95) by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

@{AndrewBarentine}  - you allude to this in the python 3 super-issue, but I can’t find a traceback. Needs more details. Effects plotting routines used in nep-fitting plugin, probably also particle tracking and PSF extraction.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

‌

```python
DEBUG:matplotlib.backends:backend WXAgg version unknown
Traceback (most recent call last):
  File "/home/smeagol/code/nep-fitting/nep_fitting/dsviewer_modules/sted_psf_fitting.py", line 172, in <lambda>
    ensemble_test_btn.Bind(wx.EVT_BUTTON, lambda e: self._on_ensemble_test())
  File "/home/smeagol/code/nep-fitting/nep_fitting/dsviewer_modules/sted_psf_fitting.py", line 369, in _on_ensemble_test
    reports.generate_and_save(fpath, context, template_name='ensemble_test.html')
  File "/home/smeagol/code/nep-fitting/nep_fitting/reports/__init__.py", line 29, in generate_and_save
    f.write(generate(context, template_name))
  File "/home/smeagol/code/nep-fitting/nep_fitting/reports/__init__.py", line 25, in generate
    return template.render(**context)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/smeagol/code/nep-fitting/nep_fitting/reports/ensemble_test.html", line 40, in top-level template code
    <td>{{ plot_data|plot(figsize=(8,6), type='scatter', xvals=ens_par, yvals=['error'], ylabel='Mean MSE [ADU^2]', c='black')}}</td>
  File "/home/smeagol/code/python-microscopy/PYME/Analysis/graphing_filters.py", line 51, in plot
    ret = mpld3.fig_to_html(f, template_type='simple')
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/site-packages/mpld3/_display.py", line 251, in fig_to_html
    figure_json=json.dumps(figure_json, cls=NumpyEncoder),
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/json/__init__.py", line 251, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/site-packages/mpld3/_display.py", line 138, in default
    return json.JSONEncoder.default(self, obj)
  File "/home/smeagol/anaconda2/envs/stablepyme/lib/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: array([1.]) is not JSON serializable
```

‌

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

Note that I have tried using tolist\(\) on both xvals and yvals, which does not fix anything. This is a mpld3 issue others have noticed, and fixed, but since the project isn’t currently maintained nothing has made it back into the head/packages.

See [here](https://stackoverflow.com/questions/47380865/json-serialization-error-using-matplotlib-mpld3-with-linkedbrush) for a discussion/solution which links a fork with fixed display.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

Also note it is not just a py3 issue, as shown in the above traceback.

‌

‌

```
(stablepyme) smeagol@smeagol:~$ conda list
# packages in environment at /home/smeagol/anaconda2/envs/stablepyme:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
alabaster                 0.7.12                   py27_0  
apptools                  4.4.0                    py27_1  
asn1crypto                0.24.0                   py27_0  
babel                     2.7.0                      py_0  
backports                 1.0                        py_2  
backports.functools_lru_cache 1.5                        py_2  
backports_abc             0.5                        py_0  
blas                      1.0                         mkl  
bokeh                     1.2.0                    py27_0  
bzip2                     1.0.6                h14c3975_5  
ca-certificates           2019.5.15                     0  
certifi                   2019.6.16                py27_1  
cffi                      1.12.3           py27h2e261b9_0  
chardet                   3.0.4                    py27_1  
cheroot                   5.8.3            py27hc34bea3_0  
cherrypy                  11.0.0           py27h05564f0_0  
click                     7.0                      py27_0  
cloudpickle               1.1.1                      py_0  
configobj                 5.0.6                    py27_1  
cryptography              2.7              py27h1ba5d50_0  
cycler                    0.10.0           py27hc7354d3_0  
cython                    0.29.12          py27he6710b0_0  
cytoolz                   0.9.0.1          py27h14c3975_1  
dask                      1.2.0                      py_0  
dask-core                 1.2.0                      py_0  
dbus                      1.13.6               h746ee38_0  
decorator                 4.4.0                    py27_1  
dispatch                  1.0                      py27_0    david_baddeley
distributed               1.28.1                   py27_0  
docutils                  0.14                     py27_0  
enum-compat               0.0.2                    py27_0    david_baddeley
enum34                    1.1.6                    py27_1  
envisage                  4.7.2                      py_0  
expat                     2.2.6                he6710b0_0  
fontconfig                2.13.0               h9420a91_0  
freetype                  2.9.1                h8a8886c_1  
functools32               3.2.3.2                  py27_1  
future                    0.17.1                   py27_0  
futures                   3.2.0                    py27_0  
glib                      2.56.2               hd408876_0  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
hdf5                      1.10.1               h9caa474_1  
heapdict                  1.0.0                    py27_2  
icu                       58.2                 h9c2bf20_1  
idna                      2.8                      py27_0  
imageio                   2.5.0                    py27_0  
imagesize                 1.1.0                    py27_0  
intel-openmp              2019.4                      243  
ipaddress                 1.0.22                   py27_0  
jaraco.functools          2.0                   py27_1000  
jinja2                    2.10.1                   py27_0  
jpeg                      9b                   h024ee3a_2  
kiwisolver                1.1.0            py27he6710b0_0  
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libpng                    1.6.37               hbc83047_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.0.10               h2733197_2  
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.13                 h1bed415_1  
libxml2                   2.9.9                he19cac6_0  
locket                    0.2.0            py27h73929a2_1  
lzo                       2.10                 h49e0be7_2  
markupsafe                1.1.1            py27h7b6447c_0  
matplotlib                2.2.3            py27hb69df0a_0  
mayavi                    4.5.0                    py27_0    david_baddeley
mkl                       2018.0.3                      1  
mkl_fft                   1.0.6            py27h7dd41cf_0  
mkl_random                1.0.1            py27h4414c95_1  
more-itertools            5.0.0                    py27_0  
mpld3                     0.3                      py27_0  
msgpack-python            0.6.1            py27hfd86e86_1  
ncurses                   6.1                  he6710b0_1  
nep-fitting               1.7                       <pip>
netifaces                 0.10.4                   py27_0    david_baddeley
networkx                  2.2                      py27_1  
numexpr                   2.6.8            py27hd89afb7_0  
numpy                     1.11.3           py27h3dfced4_4  
numpy-base                1.15.0           py27h1793315_0  
olefile                   0.46                     py27_0  
openssl                   1.1.1c               h7b6447c_1  
packaging                 19.0                     py27_0  
pandas                    0.23.4           py27h04863e7_0  
partd                     0.3.10                   py27_1  
pcre                      8.43                 he6710b0_0  
pillow                    6.0.0            py27h34e0f95_0  
pip                       19.1.1                   py27_0  
portend                   2.4                        py_0  
psutil                    5.6.2            py27h7b6447c_0  
pycparser                 2.19                     py27_0  
pyface                    6.0.0                    py27_0  
pyfftw3                   0.2.1                    py27_3    david_baddeley
pygments                  2.4.2                      py_0  
PYME                      19.8.29                   <pip>
pyme-depends              1.01                np111py27_0    david_baddeley
pymecompress              0.1.4                    py27_0    david_baddeley
pyopengl                  3.1.1a1                  py27_0  
pyopenssl                 19.0.0                   py27_0  
pyparsing                 2.4.0                      py_0  
pyqt                      5.9.2            py27h05f1152_2  
Pyro                      3.16                      <pip>
pyro                      3.16                       py_2    david_baddeley
pysocks                   1.7.0                    py27_0  
pytables                  3.4.2            py27h1f7bffc_2  
python                    2.7.16               h9bab390_0  
python-dateutil           2.8.0                    py27_0  
pytz                      2019.1                     py_0  
pywavelets                1.0.3            py27hdd07704_1  
pyyaml                    5.1              py27h7b6447c_0  
qt                        5.9.7                h5867ecd_1  
readline                  7.0                  h7b6447c_5  
requests                  2.22.0                   py27_0  
scikit-image              0.14.2           py27he6710b0_0  
scikit-learn              0.20.1           py27h4989274_0  
scipy                     1.1.0            py27hd20e5f9_0  
setuptools                41.0.1                   py27_0  
singledispatch            3.4.0.3          py27h9bcb476_0  
sip                       4.19.8           py27hf484d3e_0  
six                       1.10.0                   py27_0    david_baddeley
snowballstemmer           1.2.1            py27h44e2768_0  
sortedcontainers          2.1.0                    py27_0  
sphinx                    1.8.5                    py27_0  
sphinxcontrib             1.0                      py27_1  
sphinxcontrib-websupport  1.1.2                      py_0  
sqlite                    3.29.0               h7b6447c_0  
subprocess32              3.5.4            py27h7b6447c_0  
tblib                     1.4.0                      py_0  
tempora                   1.14.1                     py_0  
tk                        8.6.8                hbc83047_0  
toolz                     0.9.0                    py27_0  
toposort                  1.4                      py27_0    david_baddeley
tornado                   5.1.1            py27h7b6447c_0  
traits                    5.1.1            py27h7b6447c_0  
traitsui                  6.1.1                      py_0  
typing                    3.6.6                    py27_0  
ujson                     1.35             py27h14c3975_0  
urllib3                   1.24.2                   py27_0  
vtk                       6.3.0                    py27_1  
wheel                     0.33.4                   py27_0  
wxpython                  3.0.0.0                  py27_2  
xz                        5.2.4                h14c3975_4  
yaml                      0.1.7                had09818_2  
zeroconf                  0.17.8                     py_1    david_baddeley
zeroconf                  0.18.0                    <pip>
zict                      0.1.4                    py27_0  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.3.7                h0b5b093_0  
```

‌

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

This would appear to be a numpy>1.11 \(or somewhere between there and 1.14\) issue, not py3 related. Fixed by packaging most recent  version of mpld3

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Fixed by packaging (and updating to) most recent version of mpld3. OSX, Linux, and windows packages now up.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

Thanks for the fix, David! 

I ran into a small hiccup since the version number didn’t increment. I first tried a `conda update mpld3` which I thought worked, but I ran into the same error. Then I uninstalled mpld3, did a `conda clean -i`, then reinstalled. This still installed an older version from the david\_baddeley channel, namely osx-64/mpld3-0.3git-py36h1cdc3d8\_0.tar.bz2. I finally just manually downloaded the tar and installed it locally, which worked.
